lib/elixir/lib/code/typespec.ex

```diff
@@ -0,0 +1,430 @@
+    [{:->, [line: line], [args, typespec_to_quoted(result)]}]
+  end
+
+  defp typespec_to_quoted({:type, line, :fun, [args, result]}) do
+    [{:->, [line: line], [[typespec_to_quoted(args)], typespec_to_quoted(result)]}]
+  end
+
+  defp typespec_to_quoted({:type, line, :fun, []}) do
+    typespec_to_quoted({:type, line, :fun, [{:type, line, :any}, {:type, line, :any, []}]})
+  end
+
+  defp typespec_to_quoted({:type, line, :range, [left, right]}) do
+    {:.., [line: line], [typespec_to_quoted(left), typespec_to_quoted(right)]}
+  end
+
+  defp typespec_to_quoted({:type, _line, nil, []}) do
+    []
+  end
+
+  defp typespec_to_quoted({:type, line, name, args}) do
+    args = for arg <- args, do: typespec_to_quoted(arg)
+    {name, [line: line], args}
+  end
+
+  defp typespec_to_quoted({:var, line, var}) do
+    {erl_to_ex_var(var), line, nil}
+  end
+
+  defp typespec_to_quoted({:op, line, op, arg}) do
+    {op, [line: line], [typespec_to_quoted(arg)]}
+  end
+
+  defp typespec_to_quoted({:remote_type, line, [mod, name, args]}) do
+    remote_type(line, mod, name, args)
+  end
+
+  defp typespec_to_quoted({:ann_type, line, [var, type]}) do
+    {:::, [line: line], [typespec_to_quoted(var), typespec_to_quoted(type)]}
+  end
+
+  defp typespec_to_quoted(
+         {:typed_record_field, {:record_field, line, {:atom, line1, name}}, type}
+       ) do
+    typespec_to_quoted({:ann_type, line, [{:var, line1, name}, type]})
+  end
+
+  defp typespec_to_quoted({:type, _, :any}) do
+    quote(do: ...)
+  end
+
+  defp typespec_to_quoted({:paren_type, _, [type]}) do
+    typespec_to_quoted(type)
+  end
+
+  defp typespec_to_quoted({type, _line, atom}) when is_atom(type) do
+    atom
+  end
+
+  defp typespec_to_quoted(other), do: other
+
+  ## Helpers
+
+  defp remote_type(line, {:atom, _, :elixir}, {:atom, _, :charlist}, []) do
+    typespec_to_quoted({:type, line, :charlist, []})
+  end
+
+  defp remote_type(line, {:atom, _, :elixir}, {:atom, _, :nonempty_charlist}, []) do
+    typespec_to_quoted({:type, line, :nonempty_charlist, []})
+  end
+
+  defp remote_type(line, {:atom, _, :elixir}, {:atom, _, :struct}, []) do
+    typespec_to_quoted({:type, line, :struct, []})
+  end
+
+  defp remote_type(line, {:atom, _, :elixir}, {:atom, _, :as_boolean}, [arg]) do
+    typespec_to_quoted({:type, line, :as_boolean, [arg]})
+  end
+
+  defp remote_type(line, {:atom, _, :elixir}, {:atom, _, :keyword}, args) do
+    typespec_to_quoted({:type, line, :keyword, args})
+  end
+
+  defp remote_type(line, mod, name, args) do
+    args = for arg <- args, do: typespec_to_quoted(arg)
+    dot = {:., [line: line], [typespec_to_quoted(mod), typespec_to_quoted(name)]}
+    {dot, [line: line], args}
+  end
+
+  defp erl_to_ex_var(var) do
+    case Atom.to_string(var) do
+      <<"_", c::binary-1, rest::binary>> ->
+        String.to_atom("_#{String.downcase(c)}#{rest}")
+
+      <<c::binary-1, rest::binary>> ->
+        String.to_atom("#{String.downcase(c)}#{rest}")
```

Comment:
Isn't the first char the `V` we artificially add to variables? If so, we could just remove it here.

This will probably also crash on just `_` as the var name.

Reply:
Those are typed variables, so it is a whole other story. This code was just ported over but we probably need to change it to consider utf-8 variables.